odoo/orm/models.py

```diff
@@ -3355,33 +3367,108 @@ def _add_sql_constraints(self):
-                current_definition = sql.constraint_definition(cr, self._table, hashed_conname)
-                if not current_definition:
-                    _logger.info("Constraint name %r has more than 63 characters, internal PG identifier is %r", conname, hashed_conname)
-                conname = hashed_conname
-            else:
-                current_definition = sql.constraint_definition(cr, self._table, conname)
-            if current_definition == definition:
-                continue
+        for cons in self._sql_constraints:
+            cons.sync_database_object(self)
 
-            if current_definition:
-                # constraint exists but its definition may have changed
-                sql.drop_constraint(cr, self._table, conname)
+    @api.model
+    def __get_columns_from_sql_diagnostics(self, diagnostics, *, check_registry=False) -> list[str]:
+        """Given the diagnostics of an error, return the affected column names by the constraint.
+        Return an empty list if we cannot determine the columns.
+        """
+        if column := diagnostics.column_name:
+            return [column]
+        if not check_registry:
+            return []
+        cr = self.env.cr
+        cr.execute(SQL("""
+            SELECT
+                ARRAY(
+                    SELECT attname FROM pg_attribute
+                    WHERE attrelid = conrelid
+                    AND attnum = ANY(conkey)
+                ) as "columns"
+            FROM pg_constraint
+            JOIN pg_class t ON t.oid = conrelid
+            WHERE conname = %s
+                AND t.relname = %s
+        """, diagnostics.constraint_name, diagnostics.table_name))
+        [columns] = cr.fetchone() or ([])
+        return columns
 
-            if not definition:
-                # virtual constraint (e.g. implemented by a custom index)
-                self.pool.post_init(sql.check_index_exist, cr, conname)
-            elif foreign_key_re.match(definition):
-                self.pool.post_init(sql.add_constraint, cr, self._table, conname, definition)
-            else:
-                self.pool.post_constraint(sql.add_constraint, cr, self._table, conname, definition)
+    @api.model
+    def _sql_error_to_message(self, exc: psycopg2.Error) -> str:
+        """ Convert a database exception to a user error message depending on the model.
+
+        Note that the cursor on self has to be in a valid state.
+        """
+        if constraint_name := exc.diag.constraint_name:
+            for cons in self._database_objects.values():
+                if constraint_name == cons.full_name(self):
+                    cons_rec = self.env['ir.model.constraint'].search([
+                        ('name', '=', constraint_name),
+                        ('model.model', '=', self._name),
+                    ])
+                    if message := cons_rec[:1].message:
+                        return message
```

Comment:
The translation of the message actually comes from here.
We could use `limit=1` in the call to `search()` and prefetch `message` at the same time...